pkg/storage/stores/indexshipper/compactor/compactor.go

```diff
@@ -226,53 +234,76 @@ func NewCompactor(cfg Config, objectClient client.ObjectClient, schemaConfig con
 	compactor.subservicesWatcher = services.NewFailureWatcher()
 	compactor.subservicesWatcher.WatchManager(compactor.subservices)
 
-	if err := compactor.init(objectClient, schemaConfig, limits, r); err != nil {
+	if err := compactor.init(objectStoreClients, schemaConfig, limits, r); err != nil {
 		return nil, err
 	}
 
 	compactor.Service = services.NewBasicService(compactor.starting, compactor.loop, compactor.stopping)
 	return compactor, nil
 }
 
-func (c *Compactor) init(objectClient client.ObjectClient, schemaConfig config.SchemaConfig, limits *validation.Overrides, r prometheus.Registerer) error {
+func (c *Compactor) init(objectStoreClients map[string]client.ObjectClient, schemaConfig config.SchemaConfig, limits *validation.Overrides, r prometheus.Registerer) error {
 	err := chunk_util.EnsureDirectory(c.cfg.WorkingDirectory)
 	if err != nil {
 		return err
 	}
-	c.indexStorageClient = shipper_storage.NewIndexStorageClient(objectClient, c.cfg.SharedStoreKeyPrefix)
-	c.metrics = newMetrics(r)
 
 	if c.cfg.RetentionEnabled {
-		var encoder client.KeyEncoder
-		if _, ok := objectClient.(*local.FSObjectClient); ok {
-			encoder = client.FSEncoder
+		deleteRequestStore := c.cfg.DeleteRequestStore
+		// if -boltdb.shipper.compactor.shared-store is set, use it instead to ensure backward compatibility.
+		if c.cfg.SharedStoreType != "" {
```

Comment:
I think we should give precedence to `c.cfg.DeleteRequestStore` since it is a new and explicit config. Any change in that config would have been made by user deliberately. Also, it looks inconsistent not honoring an explicit config.